File: frontend-renewal/src/lib/types/metrics.ts
/**
 * Metrics types mirrored from backend Pydantic models.
 * See: backend/src/aerospike_cluster_manager_api/models/metrics.py
 */

export interface MetricPoint {
  timestamp: number;
  value: number;
}

export interface MetricSeries {
  name: string;
  label: string;
  data: MetricPoint[];
  color?: string | null;
}

export interface NamespaceMetrics {
  namespace: string;
  objects: number;
  memoryUsed: number;
  memoryTotal: number;
  deviceUsed: number;
  deviceTotal: number;
  readReqs: number;
  writeReqs: number;
  readSuccess: number;
  writeSuccess: number;
}

export interface ClusterMetrics {
  connectionId: string;
  timestamp: number;
  connected: boolean;
  uptime: number;
  clientConnections: number;
  totalReadReqs: number;
  totalWriteReqs: number;
  totalReadSuccess: number;
  totalWriteSuccess: number;
  namespaces: NamespaceMetrics[];
  readTps: MetricPoint[];
  writeTps: MetricPoint[];
  connectionHistory: MetricPoint[];
  memoryUsageByNs: MetricSeries[];
  deviceUsageByNs: MetricSeries[];
}

File: api/src/aerospike_cluster_manager_api/models/metrics.py
from __future__ import annotations

from pydantic import BaseModel


class MetricPoint(BaseModel):
    timestamp: int
    value: float


class MetricSeries(BaseModel):
    name: str
    label: str
    data: list[MetricPoint]
    color: str | None = None


class NamespaceMetrics(BaseModel):
    namespace: str
    objects: int
    memoryUsed: int
    memoryTotal: int
    deviceUsed: int
    deviceTotal: int
    readReqs: int
    writeReqs: int
    readSuccess: int
    writeSuccess: int


class ClusterMetrics(BaseModel):
    connectionId: str
    timestamp: int
    connected: bool
    uptime: int
    clientConnections: int
    totalReadReqs: int
    totalWriteReqs: int
    totalReadSuccess: int
    totalWriteSuccess: int
    namespaces: list[NamespaceMetrics]
    readTps: list[MetricPoint]
    writeTps: list[MetricPoint]
    connectionHistory: list[MetricPoint]
    memoryUsageByNs: list[MetricSeries]
    deviceUsageByNs: list[MetricSeries]
